redwood-widget/src/commonMain/kotlin/app/cash/redwood/widget/Widget.kt

```diff
@@ -69,5 +69,20 @@ public interface Widget<W : Any> {
 
     /** Indicates that [Modifier]s for the child widgets have changed. */
     public fun onModifierUpdated(index: Int, widget: Widget<W>)
+
+    /**
+     * Clear [widgets] **without** triggering an update the displayed UI.
+     *
+     * After this is called there will be no further calls to insert, remove, or move widgets, and
+     * no further updates to widgets' modifiers.
+     *
+     * Implementations must clear all direct references to [Widget] instances. This is necessary
+     * because widget implementations may be reference-counted (as in Swift) and may also contain
+     * references to Kotlin objects to create a retain cycle. We require implementations of this
+     * class to be where this cycle is broken because it's a non-invasive place to do so.
+     *
+     * This function is not recursive.
+     */
```

Comment:
I'm a bit concerned how this works (or doesn't work) if you aren't using the protocol. It seems like non-protocol widget systems should do a recursive detach, whereas protocol widget systems on the guest could no-op and leave the recursion to happen in the protocol host code, as it does now.

It's fine as-is for now, but I'm going to file an issue for the automated leak detection, and I suspect when we do that on plain Redwood usage on iOS we'd see leaks if entire subtrees were removed since only the root widget would see detach and break its reference cycle.

Reply:
Good catch. Yeah I totally missed that.

Reply:
Yeah. Future problems, though.